### PR TITLE
Rename `_SupportsStateDict` --> `_Stateful` Protocol

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Part of #11429: rename `_SupportsStateDict` --> `Stateful` Protocol
## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/types.py [hunk 1]
```diff
@@ -23,7 +23,9 @@
 from torch.optim import Optimizer
 from torch.utils.data import DataLoader
 from torchmetrics import Metric
-from typing_extensions import Protocol, runtime_checkable, TypedDict
+from typing_extensions import TypedDict
+
+from pytorch_lightning.core.hooks import Stateful
```
Comment (carmocca) [inline pytorch_lightning/utilities/types.py:28]: I strongly suggest that `Stateful` stays in this file. You'll notice there was no `from pytorch_lightning` import here and adding one opens the door for circular dependency hell, as many other source files import `types.py`

Comment (carmocca) [reply pytorch_lightning/utilities/types.py:28]: You can see this already with the red CI

Comment (ananthsub) [reply pytorch_lightning/utilities/types.py:28]: @carmocca @jjenniferdai  what if we added this to a new file: core/protocols.py ?

Comment (carmocca) [reply pytorch_lightning/utilities/types.py:28]: I don't think that would work because of the imports in `pytorch_lightning/core/__init__.py`

What's the downside of keeping it here?

What could work too is `pytorch_lightning/protocols.py`. But then we'd have to move this file too for consistency?

Comment (ananthsub) [reply pytorch_lightning/utilities/types.py:28]: I see, then moving the code around / mapping these dependencies should be handled in a separate issue/PR. The important thing now is establishing this protocol and updating the various lightning components to adopt it

**Comment by jjenniferdai** [comment]
just rename first?

File: tests/utilities/test_stateful_protocol.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Copyright The PyTorch Lightning team.
```
Comment (carmocca) [inline tests/utilities/test_stateful_protocol.py:1]: We should use a more general name, for example: `test_types.py`

**Review by carmocca — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

File: pytorch_lightning/utilities/types.py [hunk 3]
```diff
@@ -49,8 +49,8 @@
 
 
 @runtime_checkable
-class _SupportsStateDict(Protocol):
-    """This class is used to detect if an object is stateful using `isinstance(obj, _SupportsStateDict)`."""
+class Stateful(Protocol):
```
Comment (rohitgr7) [inline pytorch_lightning/utilities/types.py:52]: any reason why is it not protected anymore?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/utilities/types.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/utilities/types.py:52]: I was thinking `Stateful` protocol should be considered "public" API, same level as `CheckpointHooks`

Comment (rohitgr7) [reply pytorch_lightning/utilities/types.py:52]: can't say about `CheckpointHooks` and other hook classes now, possibly they will be made protected too since they are not used externally and are used only for the inheritance for public components like LM or LDM. But if they are targeted to be made protected too, we should make this protected now. Just to avoid 1 more possible deprecation.

cc @daniellepintz

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/utilities/types.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/utilities/types.py:52]: ah agree it's easier to elevate than deprecate so let's keep it protected for now

> else I think it will start appearing in API references in docs.

would we not want this?
